apps/rush-lib/src/logic/DependencyIntegrator.ts

```diff
@@ -0,0 +1,355 @@
+      this.updateProject(update);
+    }
+  }
+
+  /**
+   * Updates a single project's package.json file
+   */
+  public updateProject(options: IUpdateProjectOptions): void {
+    let { dependencyKind } = options;
+    const {
+      project,
+      packageName,
+      newVersion,
+      doNotSave
+    } = options;
+    const packageJson: IPackageJson = project.packageJson;
+
+    let oldDependencyKind: DependencyKind | undefined = undefined;
+    if (packageJson.dependencies && packageJson.dependencies[packageName]) {
+      oldDependencyKind = DependencyKind.Dependency;
+    } else if (packageJson.devDependencies && packageJson.devDependencies[packageName]) {
+      oldDependencyKind = DependencyKind.DevDependency;
+    }
+
+    if (!dependencyKind && !oldDependencyKind) {
+      throw new Error(`Cannot auto-detect dependency type of "${packageName}" for project "${project.packageName}"`);
+    }
+
+    if (!dependencyKind) {
+      dependencyKind = oldDependencyKind;
+    }
+
+    // update the dependency
+    if (dependencyKind === DependencyKind.Dependency) {
+      packageJson.dependencies = this._updateDependency(packageJson.dependencies, packageName, newVersion);
+    } else if (dependencyKind === DependencyKind.DevDependency) {
+      packageJson.devDependencies = this._updateDependency(packageJson.devDependencies, packageName, newVersion);
+    }
+
+    if (!doNotSave) {
+      // overwrite existing file
+      const packageJsonPath: string
+        = path.join(project.projectFolder, FileConstants.PackageJson);
+      JsonFile.save(project.packageJson, packageJsonPath);
+
+      console.log(colors.green('Wrote ') + packageJsonPath);
+    }
+  }
+
+  private _getNormalizedVersionSpec(
+    packageName: string,
+    initialSpec: string | undefined,
+    implicitlyPinnedVersion: string | undefined,
+    rangeStyle: SemVerStyle): string {
+
+    console.log(colors.gray(`Determining new version for dependency: ${packageName}`));
+    if (initialSpec) {
+      console.log(`Specified version selector: ${colors.magenta(initialSpec)}`);
+    } else {
+      console.log(`No version selector specified, will be automatically determined.`);
+    }
+    console.log();
+
+    // if ensureConsistentVersions => reuse the pinned version
+    // else, query the registry and use the latest that satisfies semver spec
+    if (initialSpec && implicitlyPinnedVersion && initialSpec === implicitlyPinnedVersion) {
+      console.log(colors.green('The specified version ')
+        + colors.magenta(initialSpec)
+        + colors.green(' has been selected as it matches the implicitly preferred version.'));
+      return initialSpec;
+    }
+
+    if (this._rushConfiguration.ensureConsistentVersions && !initialSpec && implicitlyPinnedVersion) {
+      console.log(colors.grey('The enforceConsistentVersions policy is currently active.'));
+      console.log(`Using the implicitly preferred version ${colors.magenta(implicitlyPinnedVersion)}`);
```

Comment:
>Using the implicitly preferred version ${colors.magenta(implicitlyPinnedVersion) [](start = 19, length = 80)

`Choosing the version range "^1.2.3" because it is already used by other projects in this repo.`